server/src/main/java/org/elasticsearch/search/fetch/FetchPhaseDocsIterator.java

```
@@ -113,18 +200,75 @@ public final SearchHit[] iterate(
+                chunkBuffer.clear();
+            }
+
         } catch (SearchTimeoutException e) {
             throw e;
         } catch (CircuitBreakingException e) {
             purgeSearchHits(searchHits);
+            if (streamingEnabled) {
+                purgeChunkBuffer(chunkBuffer);
+            }
             throw e;
         } catch (Exception e) {
             purgeSearchHits(searchHits);
+            if (streamingEnabled) {
+                purgeChunkBuffer(chunkBuffer);
+            }
             throw new FetchPhaseExecutionException(shardTarget, "Error running fetch phase for doc [" + currentDoc + "]", e);
         }
         return searchHits;
     }
 
+    /**
+     * Sends a chunk of hits to the coordinator.
+     */
+    private static void sendChunk(
+        FetchPhaseResponseChunk.Writer writer,
+        List<SearchHit> buffer,
+        int shardIndex,
+        int fromIndex,
+        int totalDocs,
+        float maxScore
+    ) {
+        if (buffer.isEmpty()) {
+            return;
+        }
+
+        SearchHit[] hitsArray = buffer.toArray(new SearchHit[0]);
+
+        // We incremented when adding to buffer, SearchHits constructor will increment again
+        for (SearchHit hit : hitsArray) {
+            hit.decRef();
+        }
+
+        SearchHits chunkHits = null;
+        try {
+            chunkHits = new SearchHits(hitsArray, new TotalHits(hitsArray.length, TotalHits.Relation.EQUAL_TO), maxScore);
+
+            FetchPhaseResponseChunk chunk = new FetchPhaseResponseChunk(
+                System.currentTimeMillis(),
+                FetchPhaseResponseChunk.Type.HITS,
+                shardIndex,
+                chunkHits,
+                fromIndex,
+                hitsArray.length,
+                totalDocs
+            );
+
+            writer.writeResponseChunk(chunk, ActionListener.running(() -> {}));
```

I haven't chased this through completely but `ActionListener.running(() -> {})` suggests you're sending all the chunks at once without any kind of backpressure mechanism. It's a good idea to use some amount of parallelism here but please make sure there's a (configurable?) bound on the amount of in-flight data.

---

I've updated the code to the following

`Backpressure`
We now bound the number of in-flight fetch chunks using`maxInFlightChunks`. Chunks are sent with limited parallelism, and when the bound is reached we block on the oldest in-flight request before sending more. This prevents unbounded buffering

`Failure handling`
We track the first send failure via sendFailure and fail fast: once a chunk-send fails, we stop producing further chunks and abort the fetch phase. This ensures we don’t continue doing work after a failure and that the shard fetch fails, so the coordinating node is notified.

I intentionally did not implement a retry mechanism at this stage. Retry can be added in a follow-up PR to avoid over-complicating the current change and to keep the review scope focused.

At this point, it is more important to stabilize the core behavior and sanitize test failures, which have already proven useful: they exposed a reference-count leak and highlighted the need to enforce sequence ordering when chunks arrive on the coordinating node. Addressing these correctness issues first puts us in a much better position to safely introduce retries later, if needed.

---

The main reason is that waiting for the oldest provides fail-fast behavior, if chunk 1 fails/timeouts, we detect it immediately rather than sending more chunks that will pile up. It's also simpler to reason about - don't get more than N chunks ahead.

I can see how implementing this with a semaphore approach, where we don't wait for the oldest chunk, would provide better throughput. For example, having a semaphore with (maxInFlightChunks) permits

 Permits: [✓][✓][✓]  (3 available)                                      
| -----------------------------------------------------

     *  +-----------+                                         +-------------+
     *  | Data Node |                                         | Coordinator |
     *  +-----------+                                         +-------------+
     *      |  Send Chunk 1 -acquire permit  [✓][][] -->            |
     *      |  Send Chunk 2 -acquire permit  [✓][✓][] -->           |
     *      |  Send Chunk 3 -acquire permit  [✓][✓][✓] -->          | 
     *      |                                                       |
     *      │ <-- ACK Chunk 2 arrives: [✓][✓][ ]  (release permit)  │  
     *      │  Send Chunk 4: acquire permit  [✓][✓][✓] -->          │   
     *      |                                                       |
     *      │ <-- ACK Chunk 1 arrives: [✓][✓][ ]  (release permit)  │  
     *      │  Send Chunk 5: acquire permit  [✓][✓][✓] -->          │        
     *      |                                                       |
     *      │ <-- ACK Chunk 3 arrives: [✓][✓][ ]  (release permit)  │  
     *      │  Send Chunk 6: acquire permit  [✓][✓][✓] -->          |

---

> waiting for the oldest provides fail-fast behavior, if chunk 1 fails/timeouts, we detect it immediately rather than sending more chunks that will pile up

I don't think the oldest chunk is special in that regard. In any case the coordinator is in charge of sending back the acks so it can in principle protect against a potential pile-up if that's the concern, and can do so more intelligently than the data node. That said, it's accumulating all these responses in memory anyway so a pile-up is kind of the desired outcome!